presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitSource.java

```diff
@@ -413,6 +450,91 @@ void fail(Throwable e)
+
+        ImmutableList.Builder<ConnectorSplit> result = ImmutableList.builder();
+        for (ConnectorSplit split : splits) {
+            HiveSplit hiveSplit = (HiveSplit) split;
+            List<HivePartitionKey> partitionKeys = hiveSplit.getPartitionKeys();
+            boolean removeSplit = false;
+            for (HivePartitionKey partitionKey : partitionKeys) {
+                Collection<Domain> relevantDomains = domainsCache.getDomains(partitionKey.getName());
+                if (relevantDomains != null && !relevantDomains.isEmpty()) {
+                    Type type = partitionKey.getHiveType().getType(typeManager);
+                    Domain partitionValue;
+                    try {
+                        Object objectToWrite = HiveUtil.parsePartitionValue(partitionKey.getName(), partitionKey.getValue(), type, timeZone).getValue();
+                        partitionValue = Domain.singleValue(type, objectToWrite);
+                    }
+                    catch (PrestoException e) {
+                        // if the type is not supported, skip pruning for that partition
+                        if (e.getErrorCode().equals(NOT_SUPPORTED.toErrorCode())) {
+                            log.warn("Type " + type + " is not supported for Dynamic Partition Pruning");
+                            continue;
+                        }
+
+                        throw e;
+                    }
+                    Predicate<Domain> checkOverlap = d -> d.overlaps(partitionValue);
+                    if (relevantDomains.stream().noneMatch(checkOverlap)) {
+                        log.info("Partition got pruned due to dynamic filters: "
+                                + hiveSplit.getPath() + " for table " + hiveSplit.getDatabase() + "." + hiveSplit.getTable());
+                        removeSplit = true;
+                    }
+                }
+            }
+            if (!removeSplit) {
+                result.add(split);
+            }
+        }
+        return result.build();
+    }
+
+    class DomainsCache
+    {
+        private final Map<HiveColumnHandle, Domain> domains;
+        private final Map<String, Collection<Domain>> domainsCache = new HashMap<>();
+
+        DomainsCache(Map<HiveColumnHandle, Domain> domains)
+        {
+            this.domains = ImmutableMap.copyOf(domains);
+        }
+
+        Collection<Domain> getDomains(String partitionName)
+        {
+            return domainsCache.computeIfAbsent(partitionName.toLowerCase(Locale.ENGLISH),
+                    partitionNameLower -> domains.entrySet().stream()
+                            .filter(entry -> entry.getKey().getName().equalsIgnoreCase(partitionNameLower))
+                            .map(Map.Entry::getValue)
+                            .collect(toImmutableList()));
+        }
+    }
+
+    private TupleDomain<HiveColumnHandle> getRuntimeTupleDomains()
+    {
+        // compute runtime filter
```

Comment:
Consider using a `ListenableFuture` here, so that you could do the intersection once per future (and not once per split) - similar to how it's done here:
https://github.com/prestosql/presto/blob/972c66d15f385ee05bfae1b5d5f9900471cf4466/presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java#L1979-L1994

Reply:
Good idea, I've changed the code now to intersect dynamic predicates only once per completion of future. BTW the previous approach was repeating intersections per batch of splits rather than per split.